### PR TITLE
Add support for IN predicate in outer join criteria

### PR DESCRIPTION
**Pull request by martint**

### CODE REVIEW

File: presto-main/src/test/java/io/prestosql/sql/query/TestJoin.java [hunk 1]
```diff
@@ -58,7 +59,106 @@ public void testJoinOnNan()
...
+
+        assertThat(assertions.query(
+                "SELECT * FROM (VALUES 1, 2, NULL) t(x) RIGHT JOIN (VALUES 1, 3, NULL) u(x) ON u.x IN (VALUES 1)"))
+                .matches("VALUES (1, 1), (2, 1), (NULL, 1), (NULL, 3), (NULL, NULL)");
+
+        assertThat(assertions.query(
+                "SELECT * FROM (VALUES 1, 2, NULL) t(x) FULL JOIN (VALUES 1, 3, NULL) u(x) ON u.x IN (VALUES 1)"))
+                .matches("VALUES (1, 1), (2, 1), (NULL, 1), (NULL, 3), (NULL, NULL)");
+
+        // correlation in join clause not allowed for outer join
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/sql/query/TestJoin.java:99]: what correlation in IN and inner join?

File: presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java [hunk 2]
```diff
@@ -411,15 +409,15 @@ else if (firstDependencies.stream().allMatch(right::canResolve) && secondDepende
...
-                Set<InPredicate> inPredicates = subqueryPlanner.collectInPredicateSubqueries(complexExpression, node);
-                if (!inPredicates.isEmpty()) {
-                    InPredicate inPredicate = Iterables.getLast(inPredicates);
-                    throw semanticException(NOT_SUPPORTED, inPredicate, "IN with subquery predicate in join condition is not supported");
+                Set<QualifiedName> dependencies = SymbolsExtractor.extractNames(complexExpression, analysis.getColumnReferences());
+
+                if (dependencies.stream().allMatch(left::canResolve)) {
+                    leftPlanBuilder = subqueryPlanner.handleSubqueries(leftPlanBuilder, complexExpression, node);
+                }
+                else {
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java:417]: `else if (dependencies.stream().allMatch(right::canResolve)` and then `else fail`?

Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java:417]: It doesn't matter. This is for handling uncorrelated subqueries (correlated ones are checked and prevented during analysis, as they are unsupported). This is a best effort to plan the subquery in the branch of the join that involves the operand in the expression. E.g., 

* `SELECT ... FROM t LEFT JOIN u ON t.x IN (...)`, we want the subquery to be planned in the `t` branch
* `SELECT ... FROM t LEFT JOIN u ON u.x IN (...)`, we want the subquery to be planned in the `u` branch
* `SELECT ... FROM t LEFT JOIN u ON t.x + u.x IN (...)`, handle the subquery in an arbitrary branch. Not much we can do.

I'll add a comment with an explanation.

File: presto-main/src/test/java/io/prestosql/sql/query/TestJoin.java (hunk 1 above)
Comment (martint) [reply presto-main/src/test/java/io/prestosql/sql/query/TestJoin.java:108]: Added a test.

**Comment by martint** [comment]
@kokosing, updated

File: presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java (hunk 2 above)
Comment (kokosing) [reply presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java:424]: thanks

**Review by kokosing — APPROVED** [review]

**Pull request merged**